 # **Basics Of Function**
1.max()
2.min()
3.parameter(positional,default)
4.return could be more than two values
5.doc String
6.(*) astrick
7.change the parameter
8.change the argument type
9.print() inside the loop 
10.print the function outside the loop


In [3]:
def empty_fun():
    """
       This is empty function
    """
    pass

In [9]:
list_a=[x for x in range(20)]
def max_of(list_a):
    """
      this function is used to calculat the max of a list
      
    """
    return max(list_a)
max_of(list_a)
        

19

In [11]:
price=30
def ice_cream(flavour):
    global price
    if flavour=="vanilla":
        price=price+5
    elif flavour=="choclate":
        price=price+10
    elif flavour=="strawberry":
        price=price+20
    else:
        price=price+2
    return price
ice_cream("strawberry")

50

In [2]:
tuple_a=(([x for x in range(10)]),)
def op_on_tup(tuple_a):
    return max(tuple_a)
type(op_on_tup(tuple_a))
op_on_tup(tuple_a)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [17]:
type(tuple_a)

tuple

In [20]:
tuple_b=(1,-3,4,-6,7,0)
def op_on_tup(tuple_b):
    return max(tuple_b)
type(op_on_tup(tuple_b))
op_on_tup(tuple_b)

7

In [22]:
name="Aditya"
def std(user):
    if user==name:
        print("valid user in here!!")
    elif user!=name:
        print("the invalid user is not access")
std("dfdfdf")

the invalid user is not access


In [23]:
name="Aditya"
def std(user):
    if user==name:
        print("valid user in here!!")
    elif user!=name:
        print("the invalid user is not access")
std("Aditya")

valid user in here!!


In [24]:
name="Aditya"
def std(user):
    if user=="talib":
        name="talib"
        #print("valid user in here!!")
    elif user=="Aditya":
        name="yadav"
    else:
        name="Aditya"
    return name
       # print("the invalid user is not access")
std("Aditya")

'yadav'

In [25]:
name="Aditya"
def std(user):
    if user=="talib":
        name="talib"
        #print("valid user in here!!")
    elif user=="Aditya":
        name="yadav"
    else:
        name="Aditya"
    return name
       # print("the invalid user is not access")
std("talib")

'talib'

In [26]:
price=100
def global_access(money):
    if money==price:
        print("it can only read but not write")
    else:
        print("wrong number!")
global_access(90)

wrong number!


In [27]:
global_access(50)

wrong number!


In [28]:
global_access(100)

it can only read but not write


In [3]:
price=100
def global_access(money):
    price=price+10
    if money==price:
       print("it can only read but not write")
    else:
      price=prcie-10
      print("wrong number!")
global_access(90)

UnboundLocalError: cannot access local variable 'price' where it is not associated with a value

In [39]:
price=100
def global_access(money):
    price=50
    if money==price:
        print("it can only read but not write")
    else:
        price=price-10
        print("wrong number!")
global_access(90)

wrong number!


In [29]:
price=100
def global_access(money):
    if money==price:
        price=price+10
        print("it can only read but not write")
    else:
        price=prcie-10
        print("wrong number!")
global_access(90)

UnboundLocalError: cannot access local variable 'price' where it is not associated with a value

In [30]:
name="talib"
def global_access_str(str_val):
    if str_val==name:
        print("it can only read not write")
    else:
        print("wrong string!please enter the right string")
global_access_str("aditya")

wrong string!please enter the right string


In [31]:
global_access_str("talib")

it can only read not write


In [35]:
name="talib"
def global_access_str(str_val):
    if str_val==name:
        name="ansari"
        print("it can only read not write")
    else:
        name="vipul"
        print("wrong string!please enter the right string")
global_access_str("aditya")

UnboundLocalError: cannot access local variable 'name' where it is not associated with a value

In [36]:
name="talib"
def global_access_str(str_val):
    name="ansari"
    if str_val==name:
        print("it can only read not write")
    else:
        name="vipul"
        print("wrong string!please enter the right string")
global_access_str("aditya")

wrong string!please enter the right string


In [34]:
str_name1="talib"
str_name2="ansari"
str_full_name=str_name1 + "abbu"
str_full_name

'talibabbu'

In [40]:
tuple_a=(1,2,3,4,5,6)
def max_of_tuple(tuple_a):
    return max(tuple_a)
max_of_tuple(tuple_a)

6

In [83]:
tuple_b=(1,2,3,4,5,6,7)
def index_of_tuple(tuple_b):
    if 3 in tuple_b:
        print(tuple_b.index(7))
    else:
        print("no value is there")
index_of_tuple(tuple_b)

6


In [51]:
tuple_c=(1,2,3,4,5)
tuple_d=(6,7,8,9)
def index_of_tuple(tuple_c):
    if 3 in tuple_c:
        tuple_c+=tuple_d
        print(f" tuple is changed now :{tuple_c}")
    else:
        print("no value has changed")
index_of_tuple(tuple_c)

 tuple is changed now :(1, 2, 3, 4, 5, 6, 7, 8, 9)


In [89]:
list_a=[1,2,3,4]
def global_sc_list(list_a):
    if 4 in list_a:
        list_a.append(5)
    return list_a
added_value=global_sc_list(list_a)
print(added_value)

[1, 2, 3, 4, 5]


In [60]:
list_a=[1,2,3,4]
def global_sc_list(list_a):
    if 4 in list_a:
        print(list_a.append(5))
    else:
        print("can't be changed !")
global_sc_list(list_a)

None


In [61]:
list_a

[1, 2, 3, 4, 5]

In [62]:
dict_a={
    "name":"talib",
    "age":21,
    "occupation":"student",
}
def dict_global_sc(dict_a):
    if dict_a["name"]=="talib":
        print("the same name")
    else:
        print("can't be matched")
dict_global_sc(dict_a)

the same name


In [80]:
dict_a={
    "name":"talib",
    "age":21,
    "occupation":"student",
}
def dict_global_sc(dict_a):
    if dict_a["name"]=="talib":
        dict_a.update(name="aditya",lang="python")
        print("the same name")
    else:
        print("can't be matched")
dict_global_sc(dict_a)

the same name


In [82]:
dict_a={
    "name":"talib",
    "age":21,
    "occupation":"student",
}
def dict_global_sc(dict_a):
    if dict_a["name"]=="talib":
        dict_a.update(name="aditya",lang="python")
    return dict_a
value=dict_global_sc(dict_a)
print(value)

{'name': 'aditya', 'age': 21, 'occupation': 'student', 'lang': 'python'}


In [81]:
dict_a

{'name': 'aditya', 'age': 21, 'occupation': 'student', 'lang': 'python'}

In [79]:
dict_a={
    "name":"talib",
    "age":21,
    "occupation":"student",
}
def dict_global_sc(dict_a):
    dict_a.update(lang="python")
    return dict_a
result=dict_global_sc(dict_a)
print(result)

{'name': 'talib', 'age': 21, 'occupation': 'student', 'lang': 'python'}


In [77]:
dict_a

{'name': 'talib', 'age': 21, 'occupation': 'student', 'lang': 'python'}

In [48]:
tuple_e=(1,2,3,4,5)
tuple_f=(6,7,8,9)
tuple_g=tuple_e+tuple_f
tuple_g

(1, 2, 3, 4, 5, 6, 7, 8, 9)

In [50]:
tuple_h=(1,2,3,4,5)
tuple_i=(6,7,8,9)
tuple_j=tuple_h+"talib"
tuple_j

TypeError: can only concatenate tuple (not "str") to tuple

In [1]:
tuple_i=(9,4,5,2,3)
def max_of(tuple_i):
    return max(tuple_i)
max_of(tuple_i)

9

In [2]:
def kirana_store(name,thing):
    if name=="talib" and thing=="namkeen":
        print("talib bought namkeen")
    elif name=="aditya" and thing=="biskuit":
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store("talib","namkeen")

talib bought namkeen


In [4]:
def kirana_store(name,thing):
    if name=="talib" and thing=="namkeen":
        print("talib bought namkeen")
    elif name=="aditya" and thing=="biskuit":
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store(talib,namkeen)

NameError: name 'talib' is not defined

In [6]:
def kirana_store(name,thing):
    if name==talib and thing==namkeen:
        print("talib bought namkeen")
    elif name==aditya and thing==biskuit:
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store(talib,namkeen)

NameError: name 'talib' is not defined

In [7]:
def kirana_store(name,thing):
    if name=="talib" and thing=="namkeen":
        print("talib bought namkeen")
    elif name=="aditya" and thing=="biskuit":
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store("namkeen","talib")

please enter the valid name


In [8]:
def kirana_store(name,thing):
    if name=="talib" and thing=="namkeen":
        print("talib bought namkeen")
    elif name=="aditya" and thing=="biskuit":
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store(name="talib",thing="namkeen")

talib bought namkeen


In [9]:
def kirana_store(name,thing):
    if name=="talib" and thing=="namkeen":
        print("talib bought namkeen")
    elif name=="aditya" and thing=="biskuit":
        print("aditya bought biskuit")
    else:
        print("please enter the valid name")
kirana_store(thing="namkeen",name="talib")


talib bought namkeen


In [10]:
def whole_sale(name1,name2,name3):
    name1="clothes"
    name2="utensils"
    name3="shoes"
    return name1,name2,name3
value=whole_sale(name1,name2,name3)
print(value)

NameError: name 'name1' is not defined

In [11]:
def whole_sale(name1,name2,name3):
    return clothes,shoes,bats
value=whole_sale(name1,name2,name3)
print(value)

NameError: name 'name1' is not defined

In [12]:
def whole_sale(name1,name2,name3):
    return "clothe","shoes","bats"
value=whole_sale(name1,name2,name3)
print(value)

NameError: name 'name1' is not defined

In [13]:
def whole_sale(name1,name2,name3):
    return name1,name2,name3
value=whole_sale("clothes","shoes","bats")
print(value)

('clothes', 'shoes', 'bats')


In [25]:
def whole_sale(name1,name2,name3):
    name1="talib"
    name2="bags"
    name3="bedsheet"
    return name1,name2,name3
value=whole_sale("hello","hii","hi")
print(value)

('talib', 'bags', 'bedsheet')


In [28]:
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    return name1,name2,name3,name4
shop=stationary_shop("pen","copy","sketch")
print(shop)


('pen', 'copy', 'sketch', 'tiffin_box')


In [29]:
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    return name1,name2,name3,name4
shop=stationary_shop("pen","copy")
print(shop)

TypeError: stationary_shop() missing 1 required positional argument: 'name3'

In [30]:
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    return name1,name2,name3,name4
shop=stationary_shop("pen","copy","sketch",name4="map")
print(shop)

('pen', 'copy', 'sketch', 'map')


In [31]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3):
    return name1,name2,name3
shop=stationary_shop(name1,name2,name3)
print(shop)

('pen', 'copy', 'sketch')


In [33]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3):
    name="talib"
   # return name1,name2,name3
shop=stationary_shop(name1,name2,name3)
print(shop)

None


In [34]:
stationary_shop(name1,name2,name3)

In [36]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    #name="talib"
    return name1,name2,name3,name4
shop=stationary_shop(name1,name2,name3)
print(shop)

('pen', 'copy', 'sketch', 'tiffin_box')


In [38]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    #name="talib"
    return name1,name2,name3,name4
shop=stationary_shop(name1="balck_pen",name2,name3,name4)
print(shop)

SyntaxError: positional argument follows keyword argument (692004733.py, line 7)

In [40]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    name1="black_pen"
    return name1,name2,name3,name4
shop=stationary_shop(name1,name2,name3)
print(shop)

('black_pen', 'copy', 'sketch', 'tiffin_box')


In [44]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    name1="black_pen"
    return name1,name2,name3,name4
shop=stationary_shop(name1,name2,name3,name4="lunch_box")
print(shop)

('black_pen', 'copy', 'sketch', 'lunch_box')


In [43]:
name1="pen"
name2="copy"
name3="sketch"
def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    name1="black_pen"
    return name1,name2,name3,name4
shop=stationary_shop(name1,name2,name3,name4="lunch_box",name5="rubber")
print(shop)

TypeError: stationary_shop() got an unexpected keyword argument 'name5'. Did you mean 'name1'?

In [48]:

def stationary_shop(name1,name2,name3,name4="tiffin_box"):
    name1="pen"
    name2="copy"
    name3="sketch"
    return name1,name2,name3,name4
shop=stationary_shop(name1="black_pen",name2,name3,name4="lunch_box")
print(shop)

SyntaxError: positional argument follows keyword argument (1749051230.py, line 6)

In [59]:

def stationary_shop(name1,name2,name3,name4="tiffin_box",name5):
    name1="pen"
    name2="copy"
    name3="sketch"
    name5="paper"
    return name1,name2,name3,name4,name5
shop=stationary_shop(name1,name2,name3,name5="paper_sketch")
print(shop)

SyntaxError: parameter without a default follows parameter with a default (2550033592.py, line 1)

In [64]:

def stationary_shop(name1,name2,name3,name4="tiffin_box",name5="paper"):
    name1="pen"
    name2="copy"
    name3="sketch"
    return name1,name2,name3,name4,name5
shop=stationary_shop(name1,name2,name3,name5="sketch_paper")
print(shop)

('pen', 'copy', 'sketch', 'tiffin_box', 'sketch_paper')


In [66]:

def stationary_shop(name1,name2,name3,name4="tiffin_box",name5):
    return name1,name2,name3,name4,name5
shop=stationary_shop("pen","copy","sketch","sketch_paper")
print(shop)

SyntaxError: parameter without a default follows parameter with a default (3957925550.py, line 1)

In [70]:
tuple_a=(1,2,3,4,5)
def tup_as(*a):
    return a
b=tup_as(tuple_a)
print(b)

((1, 2, 3, 4, 5),)


In [71]:
name={"talib",19,"python"}
def unpack_name(name):
    return name
a,b,c=unpack_name(name)
print(a,b,c)

talib python 19


In [73]:
a="java"
b="python"
c="C"
def pack_value(a,b,c):
    return a,b,c
d,*e=pack_value(a,b,c)
print(d,e)

java ['python', 'C']


In [75]:
a="java"
b="python"
c="C"
def pack_value(a,b,c):
    return a,b,c
e=pack_value(a,b,c)
print(*e)

java python C


In [78]:
def pack():
    return (1,2,3,4)
a=pack()
print(a)

(1, 2, 3, 4)


In [82]:
def pack():
    return (1,2,3,4)
a,*b=pack()
print(a,b)

1 [2, 3, 4]


In [83]:
def greet(name, age):
    print(f"Hello {name}, you are {age} years old.")

info = ("Alice", 30)
greet(*info)  # Unpacks tuple into name and age

details = {"name": "Bob", "age": 25}
greet(**details)  # Unpacks dictionary into keyword arguments


Hello Alice, you are 30 years old.
Hello Bob, you are 25 years old.


## Fuctions
# Built in fuction
1.ord()---->takes char to represent its ascii value
2.char()---->takes integer to represent its char value
3.abs()
4.round()
5.pow()
6.sum()
7.divmod()
8.sorted()
9.reversed()
10.filter()
11.map()
12.any()
13.all()
13.help()
14.bin()
15.hex()
16.oct()
17.eval()
18.id()
19.type()
20.min()

In [1]:
print(ord("a"))

97


In [4]:
print(ord("s"))

115


In [6]:
print(chr(65))

A


In [7]:
print(chr(127554))

🉂


In [8]:
print(chr(64674))

ﲢ


In [31]:
print(chr(128515))

😃


In [44]:
print(chr(128548))

😤


In [1]:
n=-10
abs(n)

10

In [2]:
n=27.384764
abs(n)

27.384764

In [3]:
n=-91.347
abs(n)

91.347

In [4]:
list_a=[-1,-2,-3,6.5,7,8]
abs(list_a)

TypeError: bad operand type for abs(): 'list'

In [5]:
tup_a=(-1,-2,-3,6.5,7,8)
abs(tup_a)

TypeError: bad operand type for abs(): 'tuple'

In [6]:
n=-23
round(n)

-23

In [7]:
n=46.373
round(n)

46

In [8]:
n=55.567
round(n)

56

In [9]:
def pow_of_no(n):
    return pow(n,3)
pow_of_no(3)

27

In [10]:
def add_of_num(a,b):
    return sum(a,b)
add_of_num(5,6)

TypeError: 'int' object is not iterable

In [13]:
list_a=[2,3,4,5,6]
def add_of_list(list_a):
    return sum(list_a)
add_of_list(list_a)

20

In [14]:
tuple_a=(1,2,3,-4,5,-6)
def add_of_tup(tuple_a):
    return sum(tuple_a)
add_of_tup(tuple_a)

1

In [18]:
def funtion_of_div(a,b):
    return divmod(a,b)
funtion_of_div(8,2)

(4, 0)

In [4]:
list_b=[94,6,3,48,1,23,46]
sorted(list_b)

[1, 3, 6, 23, 46, 48, 94]

In [5]:
list_b

[94, 6, 3, 48, 1, 23, 46]

In [21]:
tuple_b=(56,10,-23,-4,58,90)
sorted(tuple_b)

[-23, -4, 10, 56, 58, 90]

In [22]:
str_name="talib"
sorted(str_name)

['a', 'b', 'i', 'l', 't']

In [26]:
list_c=[x for x in range(10)]
list_c


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [27]:
reversed(list_c)

In [29]:
list_c=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
rev=reversed(list_c)
print(list(rev))

[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


In [30]:
tuple_c=(1,2,3,4,"talib","ansari")
tup=reversed(tuple_c)
print(tuple(tup))

('ansari', 'talib', 4, 3, 2, 1)


In [31]:
str_a="vipul"
str_rev=reversed(str_a)
print(''.join(str_rev))

lupiv


In [32]:
str_b="talib"
rev_str_b=reversed(str_b)
print(''.join(rev_str_b))

bilat


In [34]:
for i in reversed(range(10)):
    print(i,end=" ")

9 8 7 6 5 4 3 2 1 0 

In [37]:
def even(n):
    if n%2==0:
       return n
list_a=[1,2,3,4,56]
even_num=filter(even,list_a)
print(list(even_num))

[2, 4, 56]


In [43]:
list_a=[1,2,3,4,56]
even_num=list(filter(lambda x:x%2==0,list_a))
print(even_num)

[2, 4, 56]


In [42]:

def vowel(s):
    if s in "aeiou":
        return s
str_name="talib"
vowel_char=filter(vowel,str_name)
print(''.join(vowel_char))

ai


In [45]:
def number_check(tup_num):
    if tup_num==3:
        return tup_num
tuple_num=(91,2,3,4)
re_num=filter(number_check,tuple_num)
print(tuple(re_num))

(3,)


In [47]:
def square_num(n):
    return n**2
list_e=[1,2,3,4,5]
sq_of=map(square_num,list_e)
print(list(sq_of))

[1, 4, 9, 16, 25]


In [48]:
id(sq_of)

1876051061168

In [49]:
type(sq_of)

map

In [50]:
str_c="talib"
help(str_c)

No Python documentation found for 'talib'.
Use help() to get the interactive help utility.
Use help(str) for help on the str class.



In [51]:
str_e="choicefilling"
help(str_e)

No Python documentation found for 'choicefilling'.
Use help() to get the interactive help utility.
Use help(str) for help on the str class.



In [53]:
eval("5"+"3")

53

In [56]:
str_name1="talib"
str_name2="ansari"
eval(str_name1+str_name2)

NameError: name 'talibansari' is not defined

In [57]:
print(any([0,0,True]))

True


In [58]:
print(any([1,0,0,]))

True


In [61]:
print(any((False,True)))

True


In [62]:
print(all([1,2,3,4]))

True


In [63]:
print(all([1,0,2,3,4,4]))

False


In [64]:
bin(5)

'0b101'

In [65]:
bin(1)

'0b1'

In [66]:
hex(13)

'0xd'

In [68]:
type(hex(2635))

str

In [69]:
print(hex(2635))

0xa4b


## 30-07-2025
1.packing and unpacking
2.global and local scope
3.lambda function

In [1]:
def flavours(name):
    print("details of flavours\n")
    char_pic=chr(127848)
    print(f"{char_pic} {name}")
flavours("vanilla")

details of flavours

🍨 vanilla


In [2]:
def flavour_of_icecream(name1):
    print("details of flavours of icecream:")
    char=chr(127846)
    print(f"{char} {name1}")

flavour_of_icecream("choclate","vanilla","strawberry")  

TypeError: flavour_of_icecream() takes 1 positional argument but 3 were given

In [5]:
def flavour_of_icecream1(name1,name2,name3):
    print("details of flavours of icecream:")
    char=chr(127846)
    print(f"{char} {name1}")
    print(f"{char} {name2}")
    print(f"{char} {name3}")

flavour_of_icecream1("choclate","vanilla","strawberry") 

details of flavours of icecream:
🍦 choclate
🍦 vanilla
🍦 strawberry


In [8]:
def icecream(name1,name2,name3):
    print("Details of icecream")
    char_pic=chr(127848)
    print(f"{char_pic} {name1}")
    print(f"{char_pic} {name2}")
    print(f"{char_pic} {name3}")

icecream("choclate")

TypeError: icecream() missing 2 required positional arguments: 'name2' and 'name3'

In [28]:
def shop(name1,name2,name3):
    print("details of the things:")
    char_image=chr(129539)
    return name1
   # print(f"{char_image} {name1}")
thing=shop("bag")
print(thing)

TypeError: shop() missing 2 required positional arguments: 'name2' and 'name3'

In [27]:
def shop(name1):
    print("details of the things:")
    char_image=chr(129539)
    print(f"{char_image} {name1}")
shop("bag")
#print(thing)

details of the things:
🨃 bag


In [17]:
def shop(name1):
    print("details of the things:")
    char_image=chr(129539)
    return name1
    #print(f"{char_image} {name1}")
thing=shop("bag")
print(thing)


details of the things:
bag


In [29]:
def kirana_stores(value1,value2,value3):
    print("details of the general thing:")
    image=chr(127834)
    return value1,value2,value3
basic_things=kirana_stores("rice","wheat","dal")
print(*basic_things)
    

details of the general thing:
rice wheat dal


In [30]:
def kirana_stores(*value):
    print("details of the general thing:")
    image=chr(127834)
    return value
basic_things=kirana_stores("rice","wheat","dal")
print(basic_things)

details of the general thing:
('rice', 'wheat', 'dal')


In [31]:
def kirana_stores(*value):
    print("details of the general thing:")
    image=chr(127834)
    return value
r,w,d=kirana_stores("rice","wheat","dal")
print(r,w,d)

details of the general thing:
rice wheat dal


In [32]:
def kirana_stores(value1,value2,value3):
    print("details of the general thing:")
    image=chr(127834)
    return value1,value2,value3
r,w,d=kirana_stores("rice","wheat","dal")
print(r,w,d)

details of the general thing:
rice wheat dal


In [35]:
image=chr(127834)
char_i=chr(127822)
pic=chr(127861)
def kirana_stores(value1,value2,value3):
    print("details of the general thing:")
    print(f"{image} {value1}")
    print(f"{char_i} {value2}")
    print(f"{pic} {value3}")
kirana_stores("rice","apple","coffee")
   

details of the general thing:
🍚 rice
🍎 apple
🍵 coffee


In [38]:
image=chr(127834)
char_i=chr(127822)
pic=chr(127861)
def kirana_stores(value1,value2,value3):
    print("details of the general thing:")
    return value1,value2,value3
r,a,c=kirana_stores("rice","apple","coffee")
print(f"{image}{r}")
print(f"{char_i}{a}")
print(f"{pic}{c}")

details of the general thing:
🍚rice
🍎apple
🍵coffee


In [39]:
image=chr(127834)
char_i=chr(127822)
pic=chr(127861)
def kirana_stores(*value):
    print("details of the general thing:")
    return value
a,b,c=kirana_stores("rice","apple","coffee")
print(f"{image}{a}")
print(f"{char_i}{b}")
print(f"{pic}{c}")  

details of the general thing:
🍚rice
🍎apple
🍵coffee


In [40]:
image=chr(127834)
char_i=chr(127822)
pic=chr(127861)
def kirana_stores(*value):
    print("details of the general thing:")
    return value
a=kirana_stores("rice","apple","coffee")
print(f"{image}{a}")
print(f"{char_i}{a}")
print(f"{pic}{a}") 

details of the general thing:
🍚('rice', 'apple', 'coffee')
🍎('rice', 'apple', 'coffee')
🍵('rice', 'apple', 'coffee')


In [6]:
def clothing_shop(*clothes):
    print("details of clothes:")
    print(f"{clothes}")
    print(f"{clothes}")
clothing_shop("shirt","pant")

details of clothes:
('shirt', 'pant')
('shirt', 'pant')


In [43]:
def clothing_shop(*clothes):
    print("details of clothes here:")
    return clothes
cloth=clothing_shop("shirt","pant","coat")
print(cloth)

details of clothes here:
('shirt', 'pant', 'coat')


In [46]:
imag_of_shirt=chr(128085)
price=chr(8377)
size=32
def clothing_shop(*clothes):
    print("details of clothes here:")
    return clothes
s,p,l=clothing_shop("shirt",500,"cm")
print(f"{imag_of_shirt} {s}")
print(f"{price} {p}")
print(f"{size} {l}")

details of clothes here:
👕 shirt
₹ 500
32 cm


In [47]:
def mall(**name):
    print("details of shop in mall:")
    return name
mall(shop1="max",shop2="nike",shop3="sport")

details of shop in mall:


{'shop1': 'max', 'shop2': 'nike', 'shop3': 'sport'}

In [48]:
def mall(**name):
    print("details of shop in mall:")
    return name
mall("max",shop2="nike",shop3="sport")

TypeError: mall() takes 0 positional arguments but 1 was given

In [49]:
def production_of_mills(name,*house,**place):
    print("details of the production_of_mills:")
    return name,house,place
production_of_mills("cement","cement_production","iron_production",place1="satna",place2="shahdol")

details of the production_of_mills:


('cement',
 ('cement_production', 'iron_production'),
 {'place1': 'satna', 'place2': 'shahdol'})

In [54]:

def production_of_mills(name,*house,**place):
    print("details of the production_of_mills:")
    return name,house,place
production_of_mills("cement")

details of the production_of_mills:


('cement', (), {})

In [55]:
n=50
def scope(n):
    n=n+7
    print(n)
scope(n)

57


In [56]:
price=100
def scope1(n):
    price=price+9
    print(price)
scope1(10)

UnboundLocalError: cannot access local variable 'price' where it is not associated with a value

In [59]:
price=100
def cloths(n):
    if n=="shirt":
        price=price+500
        return price
    else:
       
       price=price+10
       return price
new_price=cloths("shirt")
print(new_price)

UnboundLocalError: cannot access local variable 'price' where it is not associated with a value

In [60]:
price=100
def cloths(n):
    global price
    if n=="shirt":
        price=price+500
        return price
    else:
       
       price=price+10
       return price
new_price=cloths("shirt")
print(new_price)

600


In [62]:
str_name="talib"
def global_scope(str_name):
    char=' '
    for i in str_name:
        char=i+char
    return char
rev_str=global_scope(str_name)
print(rev_str)
        

bilat 


In [63]:
st_name="talib"
def add_of_str(s):
    new_str=st_name+s
    return new_str
added_str=add_of_str("ansari")
print(added_str)

talibansari


In [65]:
list_num=[1,2,3,4]
def add_on_list(list_num):
    list_num.append(8)
    return list_num
add_on_list(list_num)

[1, 2, 3, 4, 8]

In [71]:
my_dict={
    "name":"mall",
    "shop":"nike",
    "product_name":"shoe"
}
def add_on_dict(s):
    if my_dict["name"]=="mall":
        my_dict.update(name="nike_mall",price=2000)
    else:
        my_dict.get("shoe_name","na")
    return my_dict
add_on_dict("mall")

{'name': 'nike_mall', 'shop': 'nike', 'product_name': 'shoe', 'price': 2000}

## Higher order funtion
1.filter()
2.map()
3.sorted()
# lambda function
variable(to store the value)=lambda input:expression

In [75]:
var=lambda x:x+x
print(var(10))

20


In [76]:
var1=lambda x:x*x
print(var1(15))

225


In [77]:
var2=lambda x:x*2
print(var2(4))

8


In [78]:
a=lambda b:b**3
print(a(3))

27


In [79]:
c=lambda d,e: d if d>e else e
print(c(10,12))

12


In [83]:
num=lambda n: "Even" if n%2==0 else "odd" 
print(num(5))

odd


In [86]:
tuple_a=(9,4,5,3,-5,6)
num=tuple(filter(lambda x: x>3,tuple_a))
print(num)

(9, 4, 5, 6)


In [90]:
list_a=[1,-3,-4,5,2,8]
num=list(map(lambda x:max(list_a),list_a))
print(num)

[8, 8, 8, 8, 8, 8]


In [91]:
list_a=[1,-3,-4,5,2,8]
num=list(filter(lambda x:x==max(list_a),list_a))
print(num)

[8]


In [92]:
list_a=[1,-3,-4,5,2,8]
num=list(map(lambda x:x==max(list_a),list_a))
print(num)

[False, False, False, False, False, True]


In [1]:
list_num=[1,2,3,4,5,6]
num=lambda rev:rev==reversed(list_num)
print(num)

<function <lambda> at 0x000001B69CB52480>


In [2]:
list_num=[1,2,3,4,5,6]
num=lambda rev:rev==reversed(list_num)
num1=list(num)
print(num1)

TypeError: 'function' object is not iterable

In [8]:
list_num=[1,2,3,4,5,6]
num=list(lambda rev:rev==reversed(list_num))
print(num)

TypeError: 'function' object is not iterable

In [4]:
list_num=[1,2,3,4,5,6]
num=list(lambda rev:rev==reversed(list_num))
num1=[]
for i in num:
    num1.append(i)
print(num1)

TypeError: 'function' object is not iterable

In [10]:
list_num=[1,2,3,4,5,6]
num=list(map(lambda list_num:reversed(list_num),list_num))
num1=[]
for i in num:
    num1.append(i)
print(num1)

TypeError: 'int' object is not reversible

In [12]:
list_num=[[1],[2],[3],[4],[5],[6]]
num=list(map(lambda rev:list(reversed(rev)),list_num))
num1=[]
for i in num:
    num1.append(i)
print(num1)

[[1], [2], [3], [4], [5], [6]]


In [13]:
add=lambda x:x+10
print(add(10))


20


In [14]:
multiplication=lambda x:x*4
print(multiplication(10))

40


In [15]:
(lambda y:y*2)(12)

24

In [16]:
(lambda x,y,z:x+y+z)(10,20,30)

60

In [18]:
difference=lambda x,y=10,z=20:x-y-z
print(difference(40))

10


In [22]:
astrick=lambda *value:sum(value)
print(astrick(10,2,30,40))

82


In [24]:
max_of=lambda *val:max(val)
print(max_of(-3,-2,-4,8,9,5,6))

9


In [25]:
min_of=lambda *val_min:min(val_min)
print(min_of(-4,-5,6,-7,8))

-7


In [31]:
list_num=[2,3,4,-8,-5]
num=list(filter(lambda x:x==min(list_num),list_num))
print(num)

[-8]


In [32]:
tup_num=(9,2,3,4,5,-7)
t_num=tuple(filter(lambda max_num:max_num==max(tup_num),tup_num))
print(t_num)

(9,)


In [35]:
add=lambda x,y:x+y()
print(add(10,lambda y=20:y+10))

40


In [36]:
product=lambda a,b:a*b(a)
print(product(10,lambda a:a*a))

1000


In [38]:
sub_string=lambda string:string in "python for Ai"
print(sub_string("python"))

True


In [39]:
str_name="pyhton is acoding language"
words=str_name.split()
for i in words:
    if i=="pyhton":
        print(" sub_string is there")
    else:
        print("sub_string is not there")
        

 sub_string is there
sub_string is not there
sub_string is not there
sub_string is not there


In [44]:
str_name="python is acoding language"
j=input()
words=str_name.split()
for i in words:
    if j==i:
        print("sub_string is there")

    else:
        print("substring is not there")
        

 python


sub_string is there
substring is not there
substring is not there
substring is not there


In [9]:
def check_sub_string(str_name):
    j=input()
    is_substring=False
    words=str_name.split()
    for i in words:
        if j==i:
            is_substring=True
            return is_substring
    return is_substring       
str_name="python is acoding language"
check_sub_string(str_name)

 pthon


False

In [68]:
sub_part=lambda string:string in "i am talib ansari"
print(sub_part("yes"))

False


In [70]:
(lambda x:(x%2 and "odd" or "even"))(10)

'even'

In [90]:
n=int(input())
add=0
for i in range(1,n):
    if n%i==0:
        add+=i
if add==n:
    print("its a perfect number")
    
else:
    print("its not a perfect number")
        

 -6


its not a perfect number


In [93]:
import math
list1=[1,2,3,4,5]
num=list(map(lambda x:pow(x,2),list1))
print(num)

[1, 4, 9, 16, 25]


In [94]:
list2=[29,30,45,67,19,78]
div_3=list(filter(lambda x:(x%3==0),list2))
print(div_3)

[30, 45, 78]


In [96]:
list2=[29,30,45,67,19,78]
list3=[]
for i in list2:
    if i%3==0:
        list3.append(i)
print(list3)
        

[30, 45, 78]


In [97]:
flavours=[("vanilla",50),("chocolate",60),("strawberry",40),("pista",90)]
price=filter(lambda flavour:flavour[1]>50,flavours)
price_fl=list(price)
print(price_fl)


[('chocolate', 60), ('pista', 90)]


In [99]:
flavours=[("vanilla",50),("chocolate",60),("strawberry",40),("pista",90)]
price=list(filter(lambda flavour:flavour[1]>50,flavours))
print(price)

[('chocolate', 60), ('pista', 90)]


In [101]:
students = (("Alice", 90), ("Bob", 75), ("Charlie", 85), ("Dave", 60))
marks=tuple(filter(lambda mark:mark[1]>75,students))
print(marks)

(('Alice', 90), ('Charlie', 85))


In [106]:
flavours_of=[("vanilla",50,True),("chocolate",60,True),("strawberry",40,False),("pista",90,False),("mango",70,True)]
availibility=list(filter(lambda available:available[2],flavours_of))
char_image=chr(127846)
r_sign=chr(8377)
for icecream,price,avail in availibility:
    print(f"{char_image} {icecream} is available : {avail}  price:{r_sign}{price}")

🍦 vanilla is available : True  price:₹50
🍦 chocolate is available : True  price:₹60
🍦 mango is available : True  price:₹70


In [113]:
icecream=[("vanilla",50),("chocolate",60),("strawberry",40),("pista",90),("mango",70)]
new_price=list(map(lambda price:(price[0],price[1]*0.18),icecream))
char_image=chr(127846)
r_sign=chr(8377)
for ice,paisa in new_price:
    print(f"{char_image}{ice}:{r_sign}{paisa:.2f}")



🍦vanilla:₹9.00
🍦chocolate:₹10.80
🍦strawberry:₹7.20
🍦pista:₹16.20
🍦mango:₹12.60


In [119]:
drinks = {
    "cola": 40,
    "juice": 55,
    "water": 20,
    "coffee": 75
}
char_image=chr(127865)
r_sign=chr(8377)
change_price=dict(map(lambda price:(price[0],price[1]+20),drinks.items()))
for i,j in change_price.items():
    print(f"{char_image}{i}:{r_sign}{j}")

🍹cola:₹60
🍹juice:₹75
🍹water:₹40
🍹coffee:₹95


In [1]:
games_dict={
    "kabaddi":8,
    "cricket":11,
    "footbal":11,
    "table_tennis":2
}
indoor_game=dict(filter(lambda game:game[1]<=2,games_dict.items()))
game_icon=chr(127918)
for key,value in indoor_game.items():
    print(f"{game_icon}{key}:{value}")

🎮table_tennis:2


In [2]:
games_dict={
    "kabaddi":8,
    "cricket":11,
    "footbal":11,
    "table_tennis":2
}
outdoor_game=dict(filter(lambda game:game[1]>2,games_dict.items()))
game_icon=chr(127918)
for key,value in outdoor_game.items():
    print(f"{game_icon}{key}:{value}")

🎮kabaddi:8
🎮cricket:11
🎮footbal:11


In [11]:
games_dict={
    "kabaddi":8,
    "cricket":11,
    "footbal":11,
    "table_tennis":2
    
}
if games_dict["kabaddi"]>7:
    games_dict["kabaddi"]=7
games_dict.update(rugby=12)
players_game=dict(filter(lambda game:game[1]>2 ,games_dict.items()))
game_icon=chr(127918)
for key,value in players_game.items():
    print(f"{game_icon}{key}:{value}")

🎮kabaddi:7
🎮cricket:11
🎮footbal:11
🎮rugby:12


In [12]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(key=lambda x:x[1],list_a)
print(sorted_list)

SyntaxError: positional argument follows keyword argument (4187214087.py, line 2)

In [13]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(list_a,key=lambda x:x[1])
print(sorted_list)

[(7, 0), (1, 2), (4, 3), (9, 5), (7, 5), (6, 7)]


In [14]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(list_a,key=lambda x:x[-1])
print(sorted_list)

[(7, 0), (1, 2), (4, 3), (9, 5), (7, 5), (6, 7)]


In [17]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(list_a,key=lambda x:x[1],reverse=True)
print(sorted_list)

[(6, 7), (9, 5), (7, 5), (4, 3), (1, 2), (7, 0)]


In [18]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(list_a,key=lambda x:x[0])
print(sorted_list)

[(1, 2), (4, 3), (6, 7), (7, 5), (7, 0), (9, 5)]


In [19]:
list_a=[(1,2),(9,5),(4,3),(7,5),(6,7),(7,0)]
sorted_list=sorted(list_a,key=lambda x:x[0],reverse=True)
print(sorted_list)

[(9, 5), (7, 5), (7, 0), (6, 7), (4, 3), (1, 2)]


In [20]:
list_str=["talib","ansari","vipul","anil"]
sort_by_char1=sorted(list_str,key=lambda x:x[0])
print(sort_by_char1)

['ansari', 'anil', 'talib', 'vipul']


In [21]:
list_str=["talib","ansari","vipul","anil"]
sort_by_char2=sorted(list_str,key=lambda x:x[1])
print(sort_by_char2)

['talib', 'vipul', 'ansari', 'anil']


In [22]:
list_str=["talib","ansari","vipul","anil"]
sort_by_char_last=sorted(list_str,key=lambda x:x[-1])
print(sort_by_char_last)

['talib', 'ansari', 'vipul', 'anil']


In [23]:
list_str=["talib","ansari","vipul","anil"]
sort_by_len=sorted(list_str,key=lambda x:len(x))
print(sort_by_len)

['anil', 'talib', 'vipul', 'ansari']


In [24]:
str_name="talib"
sort_by_ascii=sorted(str_name,key=lambda x:ascii(x))
print(sort_by_ascii)

['a', 'b', 'i', 'l', 't']


In [30]:
students = [{'name': 'Alice', 'marks': 85}, {'name': 'Bob', 'marks': 78}, {'name': 'Charlie', 'marks': 92}]
st_sort=sorted(students,key=lambda x:x["name"])
print(st_sort)

[{'name': 'Alice', 'marks': 85}, {'name': 'Bob', 'marks': 78}, {'name': 'Charlie', 'marks': 92}]


In [31]:
students = [{'name': 'Alice', 'marks': 85}, {'name': 'Bob', 'marks': 78}, {'name': 'Charlie', 'marks': 92}]
st_sort=sorted(students,key=lambda x:x["marks"])
print(st_sort)

[{'name': 'Bob', 'marks': 78}, {'name': 'Alice', 'marks': 85}, {'name': 'Charlie', 'marks': 92}]


In [33]:
students = [{'name': 'Alice', 'marks': 85}, {'name': 'Bob', 'marks': 78}, {'name': 'Charlie', 'marks': 92}]
st_sort=sorted(students,key=lambda x:x["marks"] ,reverse=True)
print(st_sort)

[{'name': 'Charlie', 'marks': 92}, {'name': 'Alice', 'marks': 85}, {'name': 'Bob', 'marks': 78}]


In [34]:
words = ['dog', 'elephant', 'cat', 'bee']
srt_by_last_char=sorted(words,key=lambda x:x[-1])
print(srt_by_last_char)

['bee', 'dog', 'elephant', 'cat']


In [35]:
list_num=[(2,3),(5,4),(9,5),(7,0),(3,9)]
sort_by_sum=sorted(list_num,key=lambda x:sum(x))
print(sort_by_sum)

[(2, 3), (7, 0), (5, 4), (3, 9), (9, 5)]


In [38]:
list_num=[(2,3),(5,4),(9,5),(7,0),(3,9)]
sort_by_product=sorted(list_num,key=lambda x:x*2)
print(sort_by_product)

[(2, 3), (3, 9), (5, 4), (7, 0), (9, 5)]


In [40]:
list_num=[(2,3),(5,4),(9,5),(7,0),(3,9)]
sort_by_sum=sorted(list_num,key=lambda x:x[0]+x[1])
print(sort_by_sum)

[(2, 3), (7, 0), (5, 4), (3, 9), (9, 5)]


## 31-07-2025 Practice Problem

1.Write a lambda function that takes two numbers and returns their sum.

In [42]:
sum_of_num=lambda x,y:x+y
print(sum_of_num(8,9))

17


In [46]:
sum_of_num=lambda x,y=9:x+y
print(sum_of_num(10))

19


In [48]:
sum_of_num=lambda x,y=9:x()+y
print(sum_of_num(x(10)))

NameError: name 'x' is not defined

2.Write a lambda function that checks whether a given number is even.

In [54]:
check_ev_od=lambda x:"odd" if x%2 else "even"
print(check_ev_od(16))

even


3.Write a lambda function that returns the square of a given number.

In [64]:
x=int(input("enter number here:"))
square_num=lambda x:x**2
print(square_num(x))


enter number here: 8


64


4.Given a list of numbers, use a lambda function with filter() to return only the positive numbers.

In [67]:
list_num=[-1,-2,-3,4,-5,-6,7,-8,9,0]
positive_num=list(filter(lambda x:x>0,list_num))
print(positive_num)

[4, 7, 9]


5.Given a list of strings, use a lambda function to sort them by their last character.

In [71]:
list_str=["hello","python","world","i","am","here","to","code"]
sort_by_last_char=sorted(list_str,key=lambda x:x[-1])
print(sort_by_last_char)

['world', 'here', 'code', 'i', 'am', 'python', 'hello', 'to']


6.Write a lambda function with map() to multiply each number in a list by 10.

In [76]:
list_tup=[("vanilla",10),("chocolate",30),("pista",40),("strawberry",50)]
increase_by_prod=list(map(lambda x:(x[0],x[1]*10),list_tup))
char_image=chr(127846)
r_sign=chr(8377)
for icecream,price in increase_by_prod:
    print(f"{char_image}{icecream} : {r_sign}{price}")


🍦vanilla : ₹100
🍦chocolate : ₹300
🍦pista : ₹400
🍦strawberry : ₹500


7.Use a lambda function to filter and count the number of vowels in a string.

In [15]:
str_name="talib"
str_vowel=''.join(filter(lambda x: x in "aeiou"  ,str_name))
print(str_vowel)

ai


In [16]:
str_name1="python is a coding language"
str_consonant=''.join(filter(lambda x:x not in "aeiou",str_name1))
print(str_consonant)

pythn s  cdng lngg


In [17]:
list_name=["talib","vipul","anil","raj"]
str_vowel=list(''.join(filter(lambda x:x[0] in"ai",list_name)))
print(str_vowel)

['a', 'n', 'i', 'l']


In [18]:
list_name=["talib","vipul","anil","raj"]
str_vowel=list(filter(lambda x:x[0] in"ai",list_name))
print(str_vowel)

['anil']


Write a lambda function that checks if a given string is a palindrome.

In [39]:
str_name2=input()
rev_str=str_name2[::-1]
palindrome=lambda str_name2,rev_str:"palindrome" if str_name2==rev_str else "not palindrome"
print(palindrome(str_name2,rev_str))

 121


palindrome


In [26]:
str_name=input()
char=''
for i in str_name:
    char=i+char
if char==str_name:
    print("it is a palindorme")
else:
    print("its not a palindrome")

    

 225522


it is a palindorme


Use a lambda function to sort a list of tuples by the second element in each tuple

In [40]:
list_tup=[("data_science",27000),("java.net",22000),("data_analytics",15000),("full_stack",30000),("dsa",40000)]
sort_by_second=sorted(list_tup,key=lambda x:x[1])
print(sort_by_second)

[('data_analytics', 15000), ('java.net', 22000), ('data_science', 27000), ('full_stack', 30000), ('dsa', 40000)]


## Problems
1.Create a function that takes a number and prints whether it is even or odd.

2.Create a function that takes two numbers and returns their sum.

3.Write a function to find the square of a number.

4.Write a function that takes a name and prints a greeting like "Hello, <name>!".

5.Write a function to calculate the factorial of a number.

6.Create a function to check if a number is prime.

7.Write a function that takes a list of numbers and returns the maximum number.

8.Write a function that takes a string and returns the number of vowels in it.

9.Write a function that accepts a string and returns it in reverse.

10.Write a function that takes a number and returns True if it is a palindrome, else False.
📌 (Palindrome: 121, 1331, 555

In [1]:
def check_ev_odd(n):
    if n % 2==0:
        print(f"{n} : is even number")
    else:
        print(f"{n} : is odd number")

In [2]:
check_ev_odd(5)

5 : is odd number


In [3]:
check_ev_odd(4)

4 : is even number


In [4]:
def check_ev_odd(n):
    if n % 2==0:
        return f"{n} is even"
    else:
        return f"{n} is odd" 
result=check_ev_odd(8)
print(result)

8 is even


In [7]:
def total_of_num(a,b):
    return a+b
result=total_of_num(7,8)
print(result)

15


In [8]:
def square_num(n):
    return n**2
square=square_num(4)
print(square)

16


In [9]:
def square_of_num(n):
    print(f"square_of {n} is:{n**2}")
square_of_num(5)

square_of 5 is:25


In [12]:
def greet(s):
    print(f"hello {s}")
greet("vipul")

hello vipul


In [22]:

def factorial(n):
    fact=1
    for i in range(n,0,-1):
        fact*=i
    return fact
 
    

In [23]:
factorial(4)

24

In [24]:
factorial(5)

120

In [25]:

def factorial(n):
    fact=1
    for i in range(n,0,-1):
        fact*=i
    print(fact)
    

In [26]:
factorial(6)

720


In [30]:
n=int(input())
def factorial(n):
    fact=1
    for i in range(n,0,-1):
        fact*=i
    return fact
factorial(n)

 7


5040

In [57]:
n=int(input())
def check_prime(n):
    if n==2:
        print("its prime")
    elif n%2==0 :
        print(f"{n} is not prime")
    else:
        print(f"{n} is prime")
check_prime(n)
    

 11


11 is prime


In [64]:
n=int(input())
def check_prime(n):
    if n<=1:
        print(f"{n} is not a prime number")
        return
    for i in range(2,(int(n**0.5)+1)):
        if n%i==0:
            print(f"{n} is not a prime number")
            return
    print(f"{n} is a prime number")
  
check_prime(n)

 -3


-3 is not a prime number


In [65]:
list_a=[x for x in range(10)]
def list_a_max(list_a):
    print(max(list_a))

In [66]:
list_a_max(list_a)

9


In [67]:
list_a=[x for x in range(10)]
def list_a_max(list_a):
    return max(list_a)
max_of_list=list_a_max(list_a)
print(max_of_list)

9


In [71]:
list_a=[x for x in range(20)]
def list_a_max(list_a):
    j=1
    for i in list_a:
        if i>j:
            j=i
    print(j)
list_a_max(list_a)

19


In [75]:
list_a=[x for x in range(20)]
def list_a_max(list_a):
    j=list_a[0]
    
    for i in list_a:
        if i>j:
            
            j=i
    print(j)
list_a_max(list_a)

19


In [74]:
list_a = [x for x in range(20)]

def list_a_max(list_a):
    j = list_a[0]  # Initialize with first element
    for i in list_a:
        if i > j:
            j = i
    print(j)

list_a_max(list_a)

19


In [83]:
s=input("Enter string :")
def vowel(s):
    count=0
    for i in s:
        if i=="a" or i=="i" or i=="e" or i=="o" or i=="u":
            
           count+=1
    return count
total=vowel(s)
print(total)
        

Enter string : ansari


3


In [85]:
s=input("Enter string :")
def vowel(s):
    count=0
    for i in s.lower():
        if i in "aeiou":
           count+=1
    return count
total=vowel(s)
print(total)

Enter string : aaaiimmmndee


7


In [89]:
a=input()
def rev_str(a):
    b=" "
    for i in a:
        b=i+b
    return b
c_str=rev_str(a)
print(c_str)

 ansari


irasna 


In [97]:
n=int(input())
def palindrome(n):
    pali=n
    rev=0
    while n>0:
        s=n%10
        rev=rev*10+s
        n//=10
    return pali==rev
rev_pali=palindrome(n)
print(rev_pali)
    

 343


True


In [113]:
str_name="talib"
str_rev=str_name[::-1]

In [114]:
str_rev

'bilat'

## 29-07-2025
1.Write a function that takes a number and prints whether it is even or odd.

In [1]:
def even_odd(n):
    if n % 2==0:
        print(f"{n} is even")
        return
    else:
        print(f"{n} is odd")
        return
even_odd(7)

7 is odd


In [6]:
n=int(input())
def even_odd(n):
    if n % 2==0:
        print(f"{n} is even")
        return
    else:
        print(f"{n} is odd")
        return
even_odd(n)

 23


23 is odd


2.Write a function that takes two numbers as arguments and returns their sum.

In [7]:
def add_of_num(a,b):
    return a+b
result=add_of_num(10,20)
print(result)

30


In [13]:
c=int(input(f"enter first number here :"))
d=int(input(f"enter second number here :"))
def add_of_num2(c,d):
    return c+d
result2=add_of_num2(c,d)
print(result2)

enter first number here : 12
enter second number here : 13


25



3.Write a function to calculate the factorial of a number.

In [17]:
def fact_of_num(n):
    fact=1
    for i in range(n,0,-1):
        fact*=i
    return fact
factorial=fact_of_num(5)
print(factorial)

120


In [23]:
n=int(input("enter the number here"))
def fact_of_num(n):
    fact=1
    for i in range(n,0,-1):
        fact*=i
    return fact
factorial=fact_of_num(n)
print(factorial)

enter the number here 19


121645100408832000


In [30]:
n=int(input("ente the number here"))
def fact_of_num(n):
    if n==1 or n==0:
        return 1
    return n*fact_of_num(n-1)
factorial=fact_of_num(n)
print(factorial)

ente the number here 5


120


4.Write a function that takes a string as input and returns its length without using len()

In [36]:
s=input()
def len_of_str(s):
    length=0
    for i in s:
        length+=1
    print(length)
len_of_str(s)

 talib


5


5.Write a function that takes a string and returns the reversed string.



In [39]:
s1=input()
def rev_str(s1):
    char=''
    for i in s1:
        char=i+char
    print(char)
rev_str(s1)

 talib


bilat


In [42]:
s1=input()
def rev_str(s1):
    for i in s1: 
        print(' '.join(i),end=" ")
rev_str(s1)

 talib


t a l i b 


6.Write a function that accepts two numbers and prints which one is greater, or if they are equal

In [46]:
a=int(input())
b=int(input())
def greater_of(a,b):
    if a>b:
        print(f"{a} is greater than{b}")
    elif a==b:
        print("both a & b are equal")
    else:
        print(f"{b} is greater than {a}")
greater_of(a,b)

 9
 0


9 is greater than0


7.Write a function that counts the number of vowels in a string.

In [50]:
str_val=input("enter the string")
def counts_vowel(str_val):
    count=0
    for i in str_val:
        if i in"aeiou":
            count+=1
    return count
counts_vowel(str_val)

enter the string eeeiiiooowwwssrr


9


8.Write a function to check whether a given string or number is a palindrome.

In [57]:
s=input()
def check_palindrome(s):
    if s.isdigit():
        n=int(s)
        pali=n
        rev=0
        while n>0:
            digit=n%10
            rev=rev*10+digit
            n=n//10
        if rev==pali:
            print("s is palindrome")
        else:
            print("it is not a palindrome")
    else:
        char=''
        for j in s:
            char=j+char
        if char==s:
            print("it is a palindrome")
        else:
            print("it is not a palindrome")
            
check_palindrome(s)

 malayalam


it is a palindrome


9.Write a function that takes a list and returns the maximum number without using max()

In [67]:
list_a=[10,2,3,11,15,18,9]
def max_of_list(list_a):
    a=list_a[0]
    for i in list_a:
        if i>a:
            a=i
    print(a)
max_of_list(list_a)

18


In [94]:
list_a=[10,2,3,11,15,18,9]

a=list_a[0]
for i in list_a:
    if i>a:
        a=i
print(a)

18


10.Write a function that prints the first n terms of the Fibonacci series

In [73]:
n=int(input())
def fibonacci(n):
    if n<=1:
        return n
    return fibonacci(n-1)+fibonacci(n-2)
for i in range(n):
    print(fibonacci(i),end=" ")
fibonacci(n)
    

 20


0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 

6765

🧠 Q11. Prime Number Checker

In [130]:
n=int(input())
def check_prime_num(n):
    if n<=1:
        print(f"{n} is not prime")
        return
    for i in range(2,(int(n**0.5)+1)):
        if n%i==0: 
          print(f"{n} is not prime")
          return
    print(f"{n} is prime")

check_prime_num(n)

 11


11 is prime


🧮 Q12. Square Each Element in List

In [132]:
list_a=[1,2,3,4,5,6,7,8,9,10]
def square_of_each_list(list_a):
    for i in list_a:
        i=i**2
        print(list(i))
square_of_each_list(list_a)

TypeError: 'int' object is not iterable

In [133]:
list_a=[1,2,3,4,5,6,7,8,9,10]
list_b=[]
def square_of_each_list(list_a):
    for i in list_a:
        i=i**2
        list_b.append(i)
        print(list_b)
square_of_each_list(list_a)

[1]
[1, 4]
[1, 4, 9]
[1, 4, 9, 16]
[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25, 36]
[1, 4, 9, 16, 25, 36, 49]
[1, 4, 9, 16, 25, 36, 49, 64]
[1, 4, 9, 16, 25, 36, 49, 64, 81]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [134]:
list_a=[1,2,3,4,5,6,7,8,9,10]
list_b=[]
def square_of_each_list(list_a):
    for i in list_a:
        i=i**2
        list_b.append(i)
    print(list_b)
square_of_each_list(list_a)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [135]:
list_a=[1,2,3,4,5,6,7,8,9,10]
list_b=[]
def square_of_each_list(list_a):
    for i in list_a:
        i=i**2
        list_b.append(i)
square_of_each_list(list_a)
print(list_b)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [136]:
list_a=[1,2,3,4,5,6,7,8,9,10]
num=list(map(lambda x:x**2,list_a))
print(num)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


13.Write a function to return the sum of digits of a number.

In [137]:
tuple_a=(9,2,4,5,6)
def sum_of_tup(tuple_a):
    return sum(tuple_a)
sum_of_tup(tuple_a)

26

In [138]:
tuple_a=(9,2,4,5,6)
def sum_of_tup(tuple_a):
    add_tup=0
    for i in tuple_a:
        add_tup+=i
    return add_tup
sum_of_tup(tuple_a)

26

In [151]:
n=int(input())
def add_of_digit(n):
    digit=0
    rem=0
    while n>0:
        rem=n%10
        digit+=rem
        n=n//10
    return digit
add_of_digit(n)

 256356335


38

In [139]:
n=1232435
p=len(n)
print(p)

TypeError: object of type 'int' has no len()

🔡 Q14. Capitalize First Letter

In [159]:
str_val="python is a coding language"
def capital_of_each_word(str_val):
    for i in str_val:
        if str_val[0].lower():
            str_val[0].capitalize()
            return 
        if i=='':
            str_val[i+1].capitalize()
            return 
    return str_val
capital_of_each_word(str_val)         

In [157]:
str_name="talib"
for i in str_name:
    if str_name[0].islower():
        str_name[0].capitalize()
    print(str_name[0])
 

t
t
t
t
t


In [155]:
str_name="talib"
str_name[0].capitalize()

'T'

In [160]:
str_val = "python is a coding language"
def capital_of_each_word(str_val):
    words = str_val.split()
    capitalized = [word.capitalize() for word in words]
    return ' '.join(capitalized)

print(capital_of_each_word(str_val))


Python Is A Coding Language


In [162]:
str_sentence="we are a modern Ai coder"
def cpaital_sent(str_sentence):
    char=str_sentence.split()
    print("character of each word:",char)
    upper_letter=[word.capitalize() for word in char]
    print("this is upper case letter:",upper_letter)
    print(' '.join(upper_letter))
cpaital_sent(str_sentence)

character of each word: ['we', 'are', 'a', 'modern', 'Ai', 'coder']
this is upper case letter: ['We', 'Are', 'A', 'Modern', 'Ai', 'Coder']
We Are A Modern Ai Coder


📊 Q15. Count Words in a Sentence

In [171]:
str_sent="we are the coder of the new modern era of the ai and python coder"
def count_words(str_sent):
    count=0
    words=str_sent.split()
    for i in range(len(words)):
        count+=i
    print(count)

count_words(str_sent )   

105


In [170]:
str_sent="we are the coder of the new modern era of the ai and python coder"
def count_words(str_sent):
    words=str_sent.split()
    
    count=len(words)
    
    print(count)

count_words(str_sent )

15


##30-07-2025 practice problem

Write a function that returns the sum of even numbers from 1 to N

In [2]:
n=int(input())
def sum_of_n(n):
    add=0
    while n>0:
        add+=n
        n-=1
    return add
sum_of_n(n)

 5


15

In [4]:
n=int(input())
def sum_of_n(n):
    add=0
    for i in range(1,n+1):
        add+=i
    return add
sum_of_n(n)

 5


15

Reverse a string using slicing inside function

In [5]:
s=input("enter the string")
def rev_s(s):
    rev_str=s[::-1]
    return rev_str
result=rev_s(s)
print(result)


enter the string vipul


lupiv


Check if a number is Armstrong number

In [19]:
import math
n=int(input())
str_s=str(n)
length=len(str_s)
def armstrong_number(n):
    add_of_num=0
    for i in str_s:
        i=int(i)
        add_of_num+=pow(i,length)
    return add_of_num
arm_strong_num=armstrong_number(n)
if arm_strong_num==n:
    print("it is a armstrong number")
else:
    print("its not an armstrong number")

        

 9474


it is a armstrong number


Return length of longest word in a sentence

str_sentence="python is a coding language"
def longest_word(str_sentence):
    words=str_sentence.split()
    for i in words:
        a=words[0]
        if i>a:
            a=words[i]
            return a
long_word=longest_word(str_sentence)
print(long_word)

In [32]:
str_sentence="python is a coding programming language"
def longest_word(str_sentence):
    words=str_sentence.split()
    a=words[0]
    for i in words:
        if len(i)>len(a):
            a=i
    return a
long_word=longest_word(str_sentence)
print(long_word)

programming


In [20]:
str_sentence="python is a coding language"
words=str_sentence.split()
print(words)

['python', 'is', 'a', 'coding', 'language']


Create a function to remove all vowels from a string

In [34]:
str_name="apeal"
def remove_vowel(str_name):
    for i in str_name:
        if i in "aeiou":
            str_name.pop(i)
    return str_name
consonant=remove_vowel(str_name)
print(consonant)
            

AttributeError: 'str' object has no attribute 'pop'

In [47]:
str_name="apeal"
def remove_vowel(str_name):
    char=' '
    for i in str_name:
        if i not in "aeiou":
            char+=i
    return char
consonant=remove_vowel(str_name)
print(consonant)

 pl


In [51]:
str_name="apeal is an art"
def remove_vowel(str_name):
    char=str_name.split()
    for i in char:
        for j in i:
            if j in "aeiou":
                char.remove(j)
    return char
        
consonant=remove_vowel(str_name)
print(consonant)

ValueError: list.remove(x): x not in list

Pass list of names → return list of names starting with 'A' (using filter function)

In [58]:
def re_names(list_a):
    list_b=[]
    for i in list_a:
        new_name="A"+i[1::]
        list_b.append(new_name)
    return list_b
list_a=["talib","vipul","brijesh"]
names=re_names(list_a)
print(names)


['Aalib', 'Aipul', 'Arijesh']


In [44]:
list_a=["talib","vipul","brijesh"]
def re_names(list_a):
    list_b=[]
    for i in list_a:
        new_name="A"+i[1::]
        list_b.append(new_name)
    print(list_b)
re_names(list_a)


['Aalib', 'Aipul', 'Arijesh']


## reduce Function ##

In [1]:
#1.first program
from functools import reduce
list_1=[1,2,3,4,5]
list_r=reduce(lambda x,y:x+y,list_1)
print(list_r)

15


In [2]:
#2.max_of
from functools import reduce
list_2=[2,7,9,-4,-5,-29,49]
list_red=reduce(lambda x,y: x if x>y else y,list_2)
print(list_red)

49


In [7]:
#3.factorial
from functools import reduce
list_3=[1,2,3,4,5,6,7]
fact_red=reduce(lambda x,y:x*y,list_3)
print(fact_red)

5040


## 1. Concatenate all strings in a list

In [8]:
from functools import reduce
list_st=["abbu","talib","ansari"]
list_concat=reduce(lambda x,y:x+y,list_st)
print(list_concat)

abbutalibansari


## 2. comput factorial

In [11]:
from functools import reduce
n=int(input("enter n here:"))
factorial=reduce(lambda x,y:x*y,range(1,n+1))
print(factorial)

enter n here: 7


5040


In [13]:
from functools import reduce
list_4=[2,1,3,4,3,5,7,2]
list_unique=reduce(lambda add,x:add+[x] if x not in add else add,list_4,[])
print(list_unique)

[2, 1, 3, 4, 5, 7]


In [14]:
from functools import reduce
list_1=[[1,2,3],[4,5,6],[7,8,9]]
flat=reduce(lambda x,y:x+y,list_1)
print(flat)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


## convert binary list to integer


In [15]:
from functools import reduce
list_binary=[1,0,1,0,1]
integer=reduce(lambda x,y:x+y,list_binary)
print(integer)

3


## custom Average

In [18]:
from functools import reduce
list_custom=[2,4,3,7,8,9]
average=reduce(lambda x,y:x+y//2,list_custom)
print(average)

16


## count vowels

In [20]:
from functools import reduce
st="abbutalibansari"
count_vowel=reduce(lambda x,y: x+1 if y in"aeiou" else x,st,0)
print(count_vowel)

7
